app.py

```diff
@@ -0,0 +1,210 @@
+# AGPL: a notification must be added stating that changes have been made to that file.
+
+import os
+
+import torch
+import torchaudio
+import streamlit as st
+
+from tortoise.api import MODELS_DIR, TextToSpeech
+from tortoise.utils.audio import load_voices
+from tortoise.utils.diffusion import K_DIFFUSION_SAMPLERS
+
+SAMPLERS = list(K_DIFFUSION_SAMPLERS.keys()) + ["ddim"]
+
+from contextlib import contextmanager
+from time import time
+from io import BytesIO
+
+
+@contextmanager
+def timeit(desc=""):
+    start = time()
+    yield
+    print(f"{desc} took {time() - start:.2f} seconds")
+
+
+if __name__ == "__main__":
+    text = st.text_area(
+        "Text",
+        help="Text to speak.",
+        value="The expressiveness of autoregressive transformers is literally nuts! I absolutely adore them.",
+    )
+
+    voices = os.listdir("tortoise/voices") + ["random"]
+    voices.remove("cond_latent_example")
+    voice = st.selectbox(
+        "Voice",
+        voices,
+        help="Selects the voice to use for generation. See options in voices/ directory (and add your own!) "
+        "Use the & character to join two voices together. Use a comma to perform inference on multiple voices.",
+        index=len(voices) - 1,
+    )
+    preset = st.selectbox(
+        "Preset",
+        (
+            "single_sample",
+            "ultra_fast",
+            "ultra_fast_old",
+            "fast",
+            "standard",
+            "high_quality",
+        ),
+        help="Which voice preset to use.",
+        index=1,
+    )
+    with st.expander("Advanced"):
+        col1, col2 = st.columns(2)
+        with col1:
+            """#### Model parameters"""
+            candidates = st.number_input(
+                "Candidates", help="How many output candidates to produce per-voice.", value=3
+            )
+            sampler = st.radio(
+                "Sampler",
+                SAMPLERS,
+                help="override the sampler used for diffusion (default depends on preset)",
+                index=1
+            )
+            steps = st.number_input(
+                "Steps",
+                help="Override the steps used for diffusion (default depends on preset)",
+                value=100,
```

Comment:
really high honestly, even DDIM doesn't need to be this high, i'll lower this unless there's a good reason not to